service/controller/azuremachinepool/handler/nodepool/delete.go

```diff
@@ -161,3 +179,39 @@ func (r *Resource) deleteVMSS(ctx context.Context, azureMachinePool *capzexpv1al
 
 	return nil
 }
+
+// removeRoleAssignment deletes the role assignment from a node pool's VMSS to allow recreation of a node pool with the same name.
+func (r *Resource) removeRoleAssignmentForPrincipalID(ctx context.Context, azureMachinePool *capzexpv1alpha3.AzureMachinePool, principalID string) error {
+	r.Logger.LogCtx(ctx, "message", "Deleting machine pool VMSS's role assignment(s)")
+
+	roleAssignmentsClient, err := r.ClientFactory.GetRoleAssignmentsClient(ctx, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	scope := fmt.Sprintf("/subscriptions/%s/resourcegroups/%s", roleAssignmentsClient.SubscriptionID, key.ClusterID(azureMachinePool))
+
+	results, err := roleAssignmentsClient.ListForScope(ctx, scope, fmt.Sprintf("principalId eq '%s'", principalID))
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.Logger.Debugf(ctx, "Found %d role assignments to delete", len(results.Values()))
```

Comment:
This deletes all role assignments found from a given resource group. How does this work when there are three NPs in a cluster and user deletes one? Are those remaining ones re-created automatically or not needed at all?

Reply:
Not true. `ListForScope` is called with a filter by principalID and the principal is unique by vmss.
Thus I'm deleting only one assignment

Reply:
![Whoa!](https://media.giphy.com/media/SDogLD4FOZMM8/giphy.gif)

Totally missed that.